libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PatternsPostProcessor.cs

```diff
@@ -0,0 +1,127 @@
+        /// </summary>
+        /// <param name="patterns">No translate patterns for different languages</param> 
+        public PatternsPostProcessor(Dictionary<string, List<string>> patterns)
+        {
+            if(patterns == null)
+            {
+                throw new ArgumentNullException(nameof(patterns));
+            }
+
+            if(patterns.Count == 0)
+            {
+                throw new ArgumentException("Patterns can't be empty");
+            }
+
+            _processedPatterns = new Dictionary<string, HashSet<string>>();
+            foreach (KeyValuePair<string, List<string>> item in patterns)
+            {
+                _processedPatterns.Add(item.Key, new HashSet<string>());
+                foreach (string pattern in item.Value)
+                {
+                    string processedLine = pattern.Trim();
+                    if (!processedLine.Contains('('))
+                    {
+                        processedLine = '(' + processedLine + ')';
+                    }
+                    _processedPatterns[item.Key].Add(processedLine);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Process the logic for patterns post processor used to handle numbers and no translate list.
+        /// </summary>
+        /// <param name="translatedDocument">Translated document</param>
+        /// <param name="currentLanguage">Current source language</param>
+        /// <returns>A Task represents the asynchronus operation</returns>
+        public PostProcessedDocument Process(TranslatedDocument translatedDocument, string currentLanguage)
+        {
+            bool containsNum = Regex.IsMatch(translatedDocument.SourceMessage, @"\d");
+            string processedResult;
+            HashSet<string> temporaryPatterns = _processedPatterns[currentLanguage];
+            if (translatedDocument.LiteranlNoTranslatePhrases.Count > 0)
+            {
+                temporaryPatterns.UnionWith((translatedDocument.LiteranlNoTranslatePhrases));
+            }
+            if (temporaryPatterns.Count == 0 && !containsNum)
+                processedResult = translatedDocument.TargetMessage;
+            if (string.IsNullOrWhiteSpace(translatedDocument.RawAlignment))
+                processedResult = translatedDocument.TargetMessage;
+
+            var toBeReplaced = from result in temporaryPatterns
+                               where Regex.IsMatch(translatedDocument.SourceMessage, result, RegexOptions.Singleline | RegexOptions.IgnoreCase)
+                               select result;
+            if (toBeReplaced.Any())
+            {
+                foreach (string pattern in toBeReplaced)
+                {
+                    Match matchNoTranslate = Regex.Match(translatedDocument.SourceMessage, pattern, RegexOptions.Singleline | RegexOptions.IgnoreCase);
+                    int noTranslateStartChrIndex = matchNoTranslate.Groups[1].Index;
```

Comment:
Comments and examples. I have no idea what this code is doing, and I don't have the time to step through it in a debugger. The more clear you can make it, with comments + examples, the better.